lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java

```diff
@@ -0,0 +1,169 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.runtime.multi.driver;
+
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.io.JsonDecoder;
+import org.apache.avro.specific.SpecificDatumReader;
+import org.apache.reef.runtime.common.driver.api.*;
+import org.apache.reef.runtime.common.driver.resourcemanager.NodeDescriptorEvent;
+import org.apache.reef.runtime.common.driver.resourcemanager.ResourceAllocationEvent;
+import org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEvent;
+import org.apache.reef.runtime.common.driver.resourcemanager.RuntimeStatusEvent;
+import org.apache.reef.runtime.multi.client.parameters.SerializedRuntimeDefinitions;
+import org.apache.reef.runtime.multi.driver.org.apache.reef.runtime.multi.driver.parameters.RuntimeName;
+import org.apache.reef.runtime.multi.utils.RuntimeDefinition;
+import org.apache.reef.tang.Configuration;
+import org.apache.reef.tang.Injector;
+import org.apache.reef.tang.JavaConfigurationBuilder;
+import org.apache.reef.tang.Tang;
+import org.apache.reef.tang.annotations.Parameter;
+import org.apache.reef.tang.exceptions.InjectionException;
+import org.apache.reef.tang.formats.AvroConfigurationSerializer;
+import org.apache.reef.wake.EventHandler;
+import org.apache.reef.wake.time.runtime.event.RuntimeStart;
+import org.apache.reef.wake.time.runtime.event.RuntimeStop;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Hosts the actual runtime implementations and delegates invocations to them.
+ */
+final class RuntimesHost {
+  private final Set<String> runtimeDefinitions;
+  private Map<String, Runtime> runtimes;
+  private final Injector originalInjector;
+  private String defaultRuntimeName;
+
+  @Inject
+  private RuntimesHost(final Injector injector,
+                       @Parameter(SerializedRuntimeDefinitions.class) final Set<String> runtimeDefinitions) {
+    if (runtimeDefinitions == null || runtimeDefinitions.size() == 0) {
+      throw new RuntimeException("No runtime configurations are provided for multi-runtime");
+    }
+
+    this.runtimeDefinitions = runtimeDefinitions;
+    this.originalInjector = injector;
+  }
+
+  private static RuntimeDefinition parseSerializedRuntimeDefinition(final String serializedRuntimeDefinition) throws
```

Comment:
This should be merged with its serialization counterpart into a class, preferably injected. That would also allow for unit tests of the (de-)serialization routines.